src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -309,68 +362,124 @@ private ObjectPropertyInitializer[] GetPropertyInitializers(IReadOnlyList<Parame
 
         protected override IEnumerable<ObjectTypeConstructor> BuildConstructors()
         {
-            yield return InitializationConstructor;
-            if (SerializationConstructor != InitializationConstructor)
+            if (!_inputModel.IsUnknownDiscriminatorModel)
+            {
+                yield return InitializationConstructor;
+            }
+
+            if (SerializationConstructor != InitializationConstructor && (!Configuration.Generation1ConvenienceClient || IncludeDeserializer))
+            {
                 yield return SerializationConstructor;
+            }
         }
 
         protected override bool EnsureHasJsonSerialization()
         {
             if (IsPropertyBag)
                 return false;
-            return true;
+            return _inputModelSerialization.Json;
         }
 
-        protected override bool EnsureHasXmlSerialization()
-        {
-            return false;
-        }
+        protected override bool EnsureHasXmlSerialization() => _inputModelSerialization.Xml is not null;
 
-        protected override bool EnsureIncludeSerializer()
-        {
-            return _inputModel.Usage.HasFlag(InputModelTypeUsage.Input);
-        }
+        protected override bool EnsureIncludeSerializer() => _inputModelUsage.HasFlag(InputModelTypeUsage.Input) && !_inputModel.IsUnknownDiscriminatorModel;
 
-        protected override bool EnsureIncludeDeserializer()
-        {
-            return _inputModel.Usage.HasFlag(InputModelTypeUsage.Output);
-        }
+        protected override bool EnsureIncludeDeserializer() => _inputModelUsage.HasFlag(InputModelTypeUsage.Output);
 
         protected override JsonObjectSerialization EnsureJsonSerialization()
         {
             // Serialization uses field and property names that first need to verified for uniqueness
             // For that, FieldDeclaration instances must be written in the main partial class before JsonObjectSerialization is created for the serialization partial class
-            return new(Type, SerializationConstructorSignature.Parameters, CreatePropertySerializations().ToArray(), null, Discriminator, false);
+            var properties = SerializationBuilder.GetPropertySerializations(this, _typeFactory);
+            var additionalProperties = CreateAdditionalPropertySerialization();
+            return new(Type, SerializationConstructor.Signature.Parameters, properties, additionalProperties, Discriminator, IncludeConverter);
         }
 
-        protected override XmlObjectSerialization? EnsureXmlSerialization()
+        private JsonAdditionalPropertiesSerialization? CreateAdditionalPropertySerialization()
         {
+            foreach (var model in EnumerateHierarchy().OfType<ModelTypeProvider>())
+            {
+                if (model is { AdditionalPropertiesProperty: {} additionalProperties, _inputModel.InheritedDictionaryType: {} inheritedDictionaryType })
```

Comment:
I believe here we only have the `{ AdditionalPropertiesProperty: {} additionalProperties }` condition is fine.
We will never have the AdditionalProperties property in the first place if the latter is false.

Reply:
It is a way to introduce `inheritedDictionaryType` variable instead of using `!` operator. Unless you want to explicitly throw when `inheritedDictionaryType` is null when `additionalProperties` is not, I'd leave it as is